diesel/src/mysql/connection/bind.rs

```diff
@@ -213,22 +218,88 @@ impl BindData {
+            MysqlType::Blob => MYSQL_TYPE_BLOB,
+            MysqlType::Numeric => MYSQL_TYPE_NEWDECIMAL,
+        }
+    }
+}
 
-    match tpe {
-        MysqlType::Tiny => MYSQL_TYPE_TINY,
-        MysqlType::Short => MYSQL_TYPE_SHORT,
-        MysqlType::Long => MYSQL_TYPE_LONG,
-        MysqlType::LongLong => MYSQL_TYPE_LONGLONG,
-        MysqlType::Float => MYSQL_TYPE_FLOAT,
-        MysqlType::Double => MYSQL_TYPE_DOUBLE,
-        MysqlType::Time => MYSQL_TYPE_TIME,
-        MysqlType::Date => MYSQL_TYPE_DATE,
-        MysqlType::DateTime => MYSQL_TYPE_DATETIME,
-        MysqlType::Timestamp => MYSQL_TYPE_TIMESTAMP,
-        MysqlType::String => MYSQL_TYPE_STRING,
-        MysqlType::Blob => MYSQL_TYPE_BLOB,
+impl From<ffi::enum_field_types> for MysqlType {
+    fn from(tpe: ffi::enum_field_types) -> Self {
+        use self::ffi::enum_field_types::*;
+
+        // https://docs.oracle.com/cd/E17952_01/mysql-8.0-en/c-api-data-structures.html
+        // https://dev.mysql.com/doc/dev/mysql-server/8.0.12/binary__log__types_8h.html
+        // https://dev.mysql.com/doc/internals/en/binary-protocol-value.html
+        // https://mariadb.com/kb/en/packet_bindata/
+        match tpe {
+            MYSQL_TYPE_TINY => MysqlType::Tiny,
+            MYSQL_TYPE_YEAR | MYSQL_TYPE_SHORT => MysqlType::Short,
+            MYSQL_TYPE_INT24 | MYSQL_TYPE_LONG => MysqlType::Long,
+            MYSQL_TYPE_LONGLONG => MysqlType::LongLong,
+            MYSQL_TYPE_FLOAT => MysqlType::Float,
+            MYSQL_TYPE_DOUBLE => MysqlType::Double,
+            MYSQL_TYPE_TIME => MysqlType::Time,
+            MYSQL_TYPE_DATE => MysqlType::Date,
+            MYSQL_TYPE_DATETIME => MysqlType::DateTime,
+            MYSQL_TYPE_TIMESTAMP => MysqlType::Timestamp,
+            MYSQL_TYPE_VAR_STRING | MYSQL_TYPE_VARCHAR | MYSQL_TYPE_STRING => MysqlType::String,
+            MYSQL_TYPE_BLOB
+            | MYSQL_TYPE_TINY_BLOB
+            | MYSQL_TYPE_MEDIUM_BLOB
+            | MYSQL_TYPE_LONG_BLOB => MysqlType::Blob,
+            MYSQL_TYPE_DECIMAL | MYSQL_TYPE_NEWDECIMAL => MysqlType::Numeric,
+            // Null value
+            MYSQL_TYPE_NULL => unreachable!(
+                "We ensure at the call side that we do not hit this type here. \
+                 If you ever see this error, something has gone very wrong. \
+                 Please open a issue at the diesel github repo in this case"
+            ),
+            // bit type
+            // same encoding as string
+            MYSQL_TYPE_BIT |
+            // enum type
+            // same encoding as string
+            MYSQL_TYPE_ENUM |
+            // set type
+            // same encoding as string
+            MYSQL_TYPE_SET |
+            // spatial type
+            // same encoding as string
+            MYSQL_TYPE_GEOMETRY |
+            // json type, only available on newer mysql versions
+            // same encoding as string
+            MYSQL_TYPE_JSON => unimplemented!(
```

Comment:
None of these appear anywhere in the docs, and it's not clear if/how we ever receive these. My guess is that they're details of the wire protocol that are handled by the client

Reply:
So I've just played a bit with various column types:
```sql
CREATE TABLE type_test (
    id INTEGER PRIMARY KEY AUTO_INCREMENT,
    some_enum ENUM ('a', 'b', 'c') NOT NULL,
    some_set SET('set_a', 'set_b', 'set_c') NOT NULL,
    jdoc JSON NOT NULL,
    geom GEOMETRY NOT NULL,
    some_bit BIT(1) NOT NULL,    
);

INSERT INTO(some_enum, some_set, jdoc, geom, some_bit) VALUES(
    'a',
    'set_b',
    '{"some_json": 42}',
     ST_GeomFromText('POINT(1 1)'),
     b'1',
);
```

Now using the code from this branch to execute the following sql queries:
```rust
    #[derive(Debug, QueryableByName)]
    struct SomeEnum {
        #[sql_type = "Text"]
        some_enum: String,
    }

    #[derive(Debug, QueryableByName)]
    struct SomeSet {
        #[sql_type = "Text"]
        some_set: String,
    }

    #[derive(Debug, QueryableByName)]
    struct Json {
        #[sql_type = "Text"]
        jdoc: String,
    }

    #[derive(Debug, QueryableByName)]
    struct Geometry {
        #[sql_type = "Text"]
        geom: String,
    }

    #[derive(Debug, QueryableByName)]
    struct Bit {
        #[sql_type = "Bool"]
        some_bit: bool,
    }

    let _ = dbg!(sql_query("SELECT some_enum FROM type_tests").get_result::<SomeEnum>(&connection));
    let _ = dbg!(sql_query("SELECT some_set FROM type_tests").get_result::<SomeSet>(&connection));
    let _ = dbg!(sql_query("SELECT jdoc FROM type_tests").get_result::<Json>(&connection));
    let _ = dbg!(sql_query("SELECT some_bit FROM type_tests").get_result::<Bit>(&connection));
    let _ = dbg!(sql_query("SELECT geom FROM type_tests").get_result::<Geometry>(&connection));
```

yields the following results:
```
[diesel/src/mysql/connection/bind.rs:251] tpe = MYSQL_TYPE_STRING
[diesel_tests/tests/types.rs:1430] sql_query("SELECT some_enum FROM type_tests").get_result::<SomeEnum>(&connection) = Ok(
    SomeEnum {
        some_enum: "b",
    },
)
[diesel/src/mysql/connection/bind.rs:251] tpe = MYSQL_TYPE_STRING
[diesel_tests/tests/types.rs:1431] sql_query("SELECT some_set FROM type_tests").get_result::<SomeSet>(&connection) = Ok(
    SomeSet {
        some_set: "set_c",
    },
)
[diesel/src/mysql/connection/bind.rs:251] tpe = MYSQL_TYPE_BLOB
[diesel_tests/tests/types.rs:1432] sql_query("SELECT jdoc FROM type_tests").get_result::<Json>(&connection) = Ok(
    Json {
        jdoc: "{\"abc\": 42, \"def\": {\"ghj\": \"foo\"}, \"baz\": []}",
    },
)
[diesel/src/mysql/connection/bind.rs:251] tpe = MYSQL_TYPE_BIT
thread 'main' panicked at 'not yet implemented: Hit currently unsupported type, those variants should probably be variants of MysqlType or at least be mapped to one of the existing types', diesel/src/mysql/connection/bind.rs:288:32
…
[diesel_tests/tests/types.rs:1434] sql_query("SELECT geom FROM type_tests").get_result::<Geometry>(&connection) = Err(
    DatabaseError(
        __Unknown,
        "Using unsupported buffer type: 255  (parameter: 1)",
    ),
)
```

This means at least the `MYSQL_TYPE_BIT` variant occurs on client side. As for the geometry error message: It seems like mys `mysqlclient` implementation itself does not support that type, probably releated to [this](https://bugs.mysql.com/bug.php?id=74315) bugreport. On the other hand the client **does** receive that type value, it just cannot handle it. (For example the native rust implementation [here](https://github.com/blackbeam/rust_mysql_common/blob/797e159514bc81eb7f88b99db0062733989045d6/src/value/mod.rs#L235) seems to handel that type.)

On the other hand doing something like this:
```rust
  table! {
        type_tests(id) {
            id -> Integer,
            some_enum -> Text,
            some_set -> Text,
            jdoc -> Text,
            geom -> Blob,
            some_bit ->Text,
        }
    }

 let dsl_res = dbg!(type_tests::table
        .select((
            type_tests::some_enum,
            type_tests::some_set,
            type_tests::jdoc,
            type_tests::geom,
            type_tests::some_bit,
        ))
        .get_result(&connection));
```

seems to work just fine:
```
[diesel_tests/tests/types.rs:1422] type_tests::table.select((type_tests::some_enum, type_tests::some_set,
                          type_tests::jdoc, type_tests::geom,
                          type_tests::some_bit)).get_result(&connection) = Ok(
    (
        "b",
        "set_c",
        "{\"abc\": 42, \"def\": {\"ghj\": \"foo\"}, \"baz\": []}",
        [
            0,
            0,
            0,
            0,
            1,
            1,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            240,
            63,
            0,
            0,
            0,
            0,
            0,
            0,
            240,
            63,
        ],
        "\u{1}",
    ),
)
```

Tested with:
* mysql  Ver 15.1 Distrib 10.3.22-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2
* libmysqlclient 8.0.19-0ubuntu0.19.10.3